lib/service/contract_base.rb

```diff
@@ -8,8 +8,29 @@ class Service::ContractBase
 
   delegate :slice, :merge, to: :to_hash
 
+  class << self
+    def attribute(name, cast_type = nil, **options, &block)
+      return super(name, cast_type, **options) unless block_given?
+
+      nested_contract_class =
+        Class.new(Service::ContractBase) do
+          define_singleton_method(:name) { "#{name}_contract".classify }
+          class_eval(&block)
+        end
+      super(
+        name,
+        Service::NestedContractType.new(
+          contract_class: nested_contract_class,
+          nested_type: cast_type || :hash,
+        ),
+        **options,
+      )
+    end
+  end
+
   def initialize(*args, options: nil, **kwargs)
     @__options__ = options
+    kwargs.deep_symbolize_keys!.slice!(*self.class.attribute_names.map(&:to_sym))
```

Comment:
The params cleaning now happens directly in the contract, that simplifies things when nesting them.